[Qt] Reduce a significant cs_main lock freeze

jonasschnelli opened the pull request:
I can't remember why we added this in the first place: but in current master, we request a `cs_main` lock (not a try lock) every block tip update in order to get the best headers height / time.
This results in significant freezes in the GUI during IBD/catch-up.

This PR adds two atomic caches for the best headers height and time.

If we do a 0.14.1rc3, we should consider adding this.

TheBlueMatt commented:
Can the GUI instead cache the current best tip? This may help with @ryanofsky's process split as well.

jonasschnelli commented:
@TheBlueMatt: Yes. @theuni recommended this on IRC and I'm currently changing it.

jonasschnelli committed:
Reduce cs_main locks during modal overlay by adding an atomic cache

jonasschnelli commented:
Moved the cache from the core validation logic to the GUI.

gmaxwell commented on src/qt/clientmodel.cpp:
```diff
@@ -72,20 +72,28 @@ int ClientModel::getNumBlocks() const
-    if (!pindexBestHeader)
-        return 0;
-    return pindexBestHeader->nHeight;
+    if (cachedBestHeaderHeight == 0) {
```
Where is this initialized? -- these don't have static duration AFAICT.

Also is zero really the best value to use? -- it's in range, so it'll keep relocking again while there is no data there.

gmaxwell commented:
What does the GUI use the header time/height for as opposed to the tip?

jonasschnelli committed:
Update the remaining blocks left in modaloverlay at init.

jonasschnelli commented:
Thanks @gmaxwell. Added the missing initialisation and switched to `-1` as indicator wether the cache is set or not.

> What does the GUI use the header time/height for as opposed to the tip?

The only use case is to calculate the remaining blocks to download/validate (modal sync progress overlay).

luke-jr commented on src/qt/clientmodel.cpp:
```diff
@@ -72,20 +74,28 @@ int ClientModel::getNumBlocks() const
 }
 
-int ClientModel::getHeaderTipHeight() const
+int ClientModel::getHeaderTipHeight()
```
Maybe this should remain const, and declare the caches mutable?

luke-jr commented on src/qt/clientmodel.cpp:
```diff
@@ -72,20 +74,28 @@ int ClientModel::getNumBlocks() const
+        // otherwise we need to wait for a tip update
+        LOCK(cs_main);
+        if (pindexBestHeader) {
+            cachedBestHeaderHeight = pindexBestHeader->nHeight;
```
Would it make sense to populate both caches at once perhaps?

jonasschnelli replied:
Good idea.

luke-jr reviewed (approved):
utACK

TheBlueMatt commented:
utACK

gmaxwell reviewed (approved):
utACK

jonasschnelli committed:
Declare headers height/time cache mutable, re-set the methods const

jonasschnelli committed:
Set both time/height header caches at the same time

jonasschnelli commented:
Followed @luke-jr's advice and re-set the method to `const` and declared the cache atomic's `mutable`.
Also, both caches are now getting set regardless of the called method. Yes, you could argue to have an explicit function to set the cache to avoid code duplication. But for two lines it's not worth in my opinion.

@gmaxwell, @TheBlueMatt: thanks for a re-ack. :)

jonasschnelli committed:
Add missing <atomic> header in clientmodel.h

laanwj commented:
Less blocking of the gui on cs_main is always a good thing, thanks!
utACK 4082fb0
